File: cpp/test/refund_transaction_entity_test.cpp
// Generated basic-flow test for the refund_transaction entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct RefundTransactionSetup {
  std::shared_ptr<BluefinTecsMerchantServicesSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static RefundTransactionSetup refund_transaction_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/refund_transaction/RefundTransactionTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinTecsMerchantServicesSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("refund_transaction01"), Value("refund_transaction02"), Value("refund_transaction03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFIN_TECS_MERCHANT_SERVICES_TEST_REFUND_TRANSACTION_ENTID", idmap},
    {"BLUEFIN_TECS_MERCHANT_SERVICES_TEST_LIVE", Value("FALSE")},
    {"BLUEFIN_TECS_MERCHANT_SERVICES_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFIN_TECS_MERCHANT_SERVICES_TEST_REFUND_TRANSACTION_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFIN_TECS_MERCHANT_SERVICES_TEST_LIVE") == Value("TRUE");

  RefundTransactionSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void refund_transaction_entity_instance() {
  auto testsdk = BluefinTecsMerchantServicesSDK::testSDK();
  auto ent = testsdk->refund_transaction();
  ASSERT_EQ(ent->getName(), std::string("refund_transaction"), "entity name");
}


static void refund_transaction_entity_basic() {
  auto setup = refund_transaction_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create"}) {
    auto sk = is_control_skipped("entityOp", std::string("refund_transaction.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto refund_transaction_ref01_ent = client->refund_transaction();
  Value refund_transaction_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "refund_transaction"}), "refund_transaction_ref01"));
  if (!refund_transaction_ref01_data.is_map()) refund_transaction_ref01_data = vmap();
  {
    Value refund_transaction_ref01_data_result = refund_transaction_ref01_ent->create(Struct::clone(refund_transaction_ref01_data), Value::undef())->data();
    refund_transaction_ref01_data = Helpers::toMapAny(refund_transaction_ref01_data_result);
    if (!refund_transaction_ref01_data.is_map()) refund_transaction_ref01_data = vmap();
    ASSERT_TRUE(refund_transaction_ref01_data.is_map(), "expected create result to be a map");
  }

}

int main() {
  T_RUN(refund_transaction_entity_instance);
  T_RUN(refund_transaction_entity_basic);
  return sdktest::summary("refund_transaction_entity_test");
}

File: scala/sdktest/RefundTransactionEntityTest.scala
// Generated basic-flow test for the refund_transaction entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped RefundTransactionTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.bluefintecsmerchantservicessdk.core.{Helpers, SdkEntity, BluefinTecsMerchantServicesSDK}
import voxgig.bluefintecsmerchantservicessdk.utility.struct.Struct

object RefundTransactionEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("refund_transaction.instance") {
      val testsdk = BluefinTecsMerchantServicesSDK.testSDK()
      val ent = testsdk.refundTransaction(null)
      rep.check("refund_transaction.instance", ent != null, "expected non-null refund_transaction entity")
    }

    rep.scope("refund_transaction.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/refund_transaction/RefundTransactionTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = BluefinTecsMerchantServicesSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("refund_transaction01", "REFUND_TRANSACTION01")
      idmap.put("refund_transaction02", "REFUND_TRANSACTION02")
      idmap.put("refund_transaction03", "REFUND_TRANSACTION03")
      val now = System.currentTimeMillis()

      // CREATE
      val refundTransactionRef01Ent = client.refundTransaction(null)
      var refundTransactionRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.refund_transaction"), "refund_transaction_ref01"))
      val refundTransactionRef01DataResult = refundTransactionRef01Ent.create(refundTransactionRef01Data, null)
      refundTransactionRef01Data = Helpers.toMapAny(refundTransactionRef01DataResult match { case e: SdkEntity => e.data(); case o => o })
      rep.check("refund_transaction.create.map", refundTransactionRef01Data != null, "expected create result to be a map")
    }
  }
}
